File: java/src/Matrix.java
import java.util.Random;

public class Matrix {
    /**
     * Matrix multiplication implementation.
     */
    public static double[][] matrixMultiply(double[][] matrixA, double[][] matrixB) {
        int rowsA = matrixA.length;
        int colsA = matrixA[0].length;
        int colsB = matrixB[0].length;

        // Initialize result matrix with zeros
        double[][] result = new double[rowsA][colsB];

        // Perform multiplication
        for (int i = 0; i < rowsA; i++) {
            for (int j = 0; j < colsB; j++) {
                for (int k = 0; k < colsA; k++) {
                    result[i][j] += matrixA[i][k] * matrixB[k][j];
                }
            }
        }

        return result;
    }

    /**
     * Run matrix multiplication benchmark with 100x100 matrices.
     */
    public static void runBenchmark() {
        int size = 100;
        Random rand = new Random();

        // Generate two random 100x100 matrices
        double[][] matrixA = new double[size][size];
        double[][] matrixB = new double[size][size];
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < size; j++) {
                matrixA[i][j] = rand.nextDouble();
                matrixB[i][j] = rand.nextDouble();
            }
        }

        // Measure multiplication time
        long startTime = System.nanoTime();
        double[][] result = matrixMultiply(matrixA, matrixB);
        long endTime = System.nanoTime();
        long executionTime = (endTime - startTime) / 1000000;

        System.out.println("Test: Matrix Multiplication");
        System.out.println("Matrix size: " + size + "x" + size);
        System.out.println("Execution time: " + executionTime + " ms");
        System.out.printf("Result sample (0,0): %.6f%n", result[0][0]);
    }

    public static void main(String[] args) {
        runBenchmark();
    }
}

File: ruby/matrix.rb
require 'benchmark'

# Matrix multiplication implementation
def matrix_multiply(matrix_a, matrix_b)
  rows_a = matrix_a.length
  cols_a = matrix_a[0].length
  cols_b = matrix_b[0].length

  # Initialize result matrix with zeros
  result = Array.new(rows_a) { Array.new(cols_b, 0.0) }

  # Perform multiplication
  (0...rows_a).each do |i|
    (0...cols_b).each do |j|
      (0...cols_a).each do |k|
        result[i][j] += matrix_a[i][k] * matrix_b[k][j]
      end
    end
  end

  result
end

# Run matrix multiplication benchmark with 100x100 matrices
def run_benchmark
  size = 100

  # Generate two random 100x100 matrices
  matrix_a = Array.new(size) { Array.new(size) { rand } }
  matrix_b = Array.new(size) { Array.new(size) { rand } }

  # Measure multiplication time
  start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)
  result = matrix_multiply(matrix_a, matrix_b)
  end_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)
  execution_time = ((end_time - start_time) * 1000).round(2)

  puts "Test: Matrix Multiplication"
  puts "Matrix size: #{size}x#{size}"
  puts "Execution time: #{execution_time} ms"
  puts "Result sample (0,0): #{result[0][0].round(6)}"
end

if __FILE__ == $0
  run_benchmark
end
